src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -805,6 +805,162 @@ case class TableDistinct(child: TableIR) extends TableIR {
+  }
+
+  private val keyType = newKey.typ.asInstanceOf[TStruct]
+  val typ: TableType = TableType(rowType = keyType ++ coerce[TStruct](expr.typ),
+    globalType = child.typ.globalType,
+    key = Some(keyType.fieldNames)
+  )
+
+  def execute(hc: HailContext): TableValue = {
+    val prev = child.execute(hc)
+
+    val (rvAggs, makeInit, makeSeq, aggResultType, postAggIR) = ir.CompileWithAggregators[Long, Long, Long](
+      "global", child.typ.globalType,
+      "global", child.typ.globalType,
+      "row", child.typ.rowType,
+      expr, "AGGR",
+      (nAggs, initializeIR) => initializeIR,
+      (nAggs, sequenceIR) => sequenceIR)
+
+    val (rTyp, makeAnnotate) = ir.Compile[Long, Long, Long](
+      "AGGR", aggResultType,
+      "global", child.typ.globalType,
+      postAggIR)
+
+    val nAggs = rvAggs.length
+
+    assert(coerce[TStruct](rTyp) == typ.valueType, s"$rTyp, ${ typ.valueType }")
+
+    val rowType = prev.typ.rowType
+    val globalsType = prev.typ.globalType
+    val globalsBc = prev.globals.broadcast
+
+    val localKeyType = keyType
+    val newValueType = typ.valueType
+    val newRowType = typ.rowType
+    val (_, makeKeyF) = ir.Compile[Long, Long, Long](
+      "row", child.typ.rowType,
+      "global", child.typ.globalType,
+      newKey
+    )
+
+    val localBufferSize = bufferSize
+    val combOp = { (aggs1: Array[RegionValueAggregator], aggs2: Array[RegionValueAggregator]) =>
+      var i = 0
+      while (i < aggs2.length) {
+        aggs1(i).combOp(aggs2(i))
+        i += 1
+      }
+      aggs1
+    }
+
+    val rdd = try {
+      prev.rvd
+        .boundary
+        .mapPartitions { (ctx, it) =>
+          val rvb = new RegionValueBuilder()
+          val partRegion = ctx.freshContext.region
+
+          rvb.set(partRegion)
+          rvb.start(globalsType)
+          rvb.addAnnotation(globalsType, globalsBc.value)
+          val globals = rvb.end()
+          val init = makeInit()
+          init(partRegion, rvAggs, globals, false)
+
+          val makeKey = {
+            val f = makeKeyF()
+            rv: RegionValue => {
+              val keyOff = f(rv.region, rv.offset, false, globals, false)
+              SafeRow.read(localKeyType, rv.region, keyOff).asInstanceOf[Row]
+            }
+          }
+          val sequence = {
+            val f = makeSeq()
+            (rv: RegionValue, rvAggs: Array[RegionValueAggregator]) => {
+              f(rv.region, rvAggs, globals, false, rv.offset, false)
+            }
+          }
+          new BufferedAggregatorIterator[RegionValue, Array[RegionValueAggregator], Row](
+            it,
+            () => rvAggs.map(_.copy()),
+            makeKey,
+            sequence,
+            localBufferSize)
+        }.reduceByKey(combOp, nPartitions.getOrElse(prev.rvd.getNumPartitions))
+    } catch {
+      // support empty tables
+      case u: UnsupportedOperationException => hc.sc.emptyRDD[(Row, Array[RegionValueAggregator])]
```

Comment:
I prefer aggregateByKey here instead.  This can potentially mask other bugs that throw unsupported operation.

Reply:
I originally had this, but that needs a zero value, and the logic to initialize it is a bit nasty. Should I make a RV for globals locally, init `rvAggs`, and then `aggregateByKey`? This seemed simpler.

Reply:
done.

Reply:
I was worried about the region manipulation on driver, but Region.scoped was exactly what I needed.